### PR TITLE
More clean up for RegexOptions.NonBacktracking

### PR DESCRIPTION
**Pull request by stephentoub**
There's still a lot more to do, but this is as far as I got today between meetings...

- Added more comments
- Removed some dead code and ifdef'd some code to DEBUG-only
- Updated an IComparable implementation to IComparable<T>
- Sealed some internal types
- Made UnicodeCategoryTheory into a static type
- Made CharSetSolver into a singleton
- Renamed a bunch of Mk* methods to Create*
- Renamed Watchdog
- Removed duplicative helper functions
- Fixed thread-safety of UnicodeCategoryTheory (in practice it was probably fine but we now have true singletons)
- Changed some iterators to be lists
- Used var less
- Cleaned up the set parsing code in RegexNodeConverter
- Lazily-initialized a bunch of members of SymbolicRegexBuilder
- Fixed a bug in the handling of \w
- And some other style stuff...

cc: @joperezr, @olsaarik

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeToSymbolicConverter.cs [hunk 1]
```diff
@@ -5,326 +5,194 @@
...
+                case RegexNodeKind.Onelazy:
+                case RegexNodeKind.Notoneloop:
+                case RegexNodeKind.Notonelazy:
                     {
-                        catCodes.Remove(0);
-                        catCodes.Remove(1);
-                        catCodes.Remove(2);
-                        catCodes.Remove(3);
-                        catCodes.Remove(4);
-                        catCodes.Remove(8);
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeToSymbolicConverter.cs:184]: @olsaarik, this was a bug that it was missing a Contains/Remove for 5, yes?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Unicode/UnicodeCategoryTheory.cs [hunk 2]
```diff
@@ -1,77 +1,75 @@
...
 
+using System.Diagnostics;
+using System.Globalization;
+using System.Threading;
+
 namespace System.Text.RegularExpressions.Symbolic.Unicode
 {
-    internal sealed class UnicodeCategoryTheory<TPredicate> where TPredicate : class
+    /// <summary>Utility class providing singleton <see cref="BDD"/>s for evaluating whether a character is part of a particular Unicode category.</summary>
+    internal static class UnicodeCategoryConditions
```
Comment (teo-tsirpanis) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Unicode/UnicodeCategoryTheory.cs:11]: You could rename the file accordingly.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Unicode/UnicodeCategoryTheory.cs:11]: I plan to subsequently. It can be hard to follow diffs in my experience when there are lots of changes in a file and it's renamed.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs [hunk 3]
```diff
@@ -30,19 +32,19 @@ internal DfaMatchingState(SymbolicRegexNode<T> node, uint prevCharKind)
         internal bool IsDeadend => Node.IsNothing;
 
         /// <summary>The node must be nullable here</summary>
-        internal int WatchDog
+        internal int FixedLength
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs:35]: NIT: Adjust triple slash docs as well.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs [hunk 4]
```diff
@@ -106,20 +108,20 @@ internal DfaMatchingState<T> Next(T minterm)
...
             // otherwise create a new new id for it
-            return Node._builder.MkState(derivative, nextCharKind, capturing: false);
+            return Node._builder.CreateState(derivative, nextCharKind, capturing: false);
         }
 
         /// <summary>
         /// Compute a set of transitions for the given minterm.
         /// </summary>
         /// <param name="minterm">minterm corresponding to some input character or False corresponding to last \n</param>
         /// <returns>an enumeration of the transitions as pairs of the target state and a list of effects to be applied</returns>
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs:123]:
```suggestion
        /// <returns>a list of the transitions as pairs of the target state (derivative) and a list of effects to be applied</returns>
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+    {
+        /// <summary>The culture to use for IgnoreCase comparisons.</summary>
+        private readonly CultureInfo _culture;
+        /// <summary>Capture information.</summary>
+        private readonly Hashtable? _caps;
+        /// <summary>The builder to use to create the <see cref="SymbolicRegexNode{S}"/> nodes.</summary>
+        internal readonly SymbolicRegexBuilder<BDD> _builder;
+
+        /// <summary>Cache of BDDs created to represent <see cref="RegexCharClass"/> set strings.</summary>
+        /// <remarks>This cache is useful iff the same character class is used multiple times in the same regex, but that's fairly common.</remarks>
```
[the remark this answers is not in the document]
Comment (joperezr) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs:25]: also, I asume the `but that's fairly common` -> `which is fairly common`?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeToSymbolicConverter.cs (hunk 1 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeToSymbolicConverter.cs:184]: Yes, it is indeed a bug, because \w is the union of the 8 categories: 0,1,2,3,4,5,8,18

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs (hunk 3 above)
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs:35]: The renaming of `WatchDog` will cause merge conflict with my recent PR, but that should be fine. I may ask help with merging ... or someone else may merge :-)

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (veanes) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs:25]: I think that the common thing is that the same character class, such as say [@#-;] may be used multiple times in the same regex.

**Review by veanes — APPROVED** [review]
I agree with the changes, but many of the changes will cause direct naming conflicts all over with my recent PR   #65637 
So **quite a bit** of hand-editing will be needed, whoever merges it in at that point, essentially naming changes in method/field names.

**Comment by stephentoub** [comment]
> whoever merges it in at that point

I'll take care of it.
